src/linux/osd.c

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright (c) 2017 Intel Corp., Inc.  All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <pthread.h>
+
+#include <fi_osd.h>
+
+int ofi_set_thread_affinity(const char *s)
+{
+	char *saveptra = NULL, *saveptrb = NULL, *saveptrc = NULL;
+	char *dup_s, *a, *b, *c;
+	int j, first, last, stride, ret = FI_SUCCESS;
+	cpu_set_t mycpuset;
+	pthread_t mythread;
+
+	mythread = pthread_self();
+	CPU_ZERO(&mycpuset);
+
+	dup_s = strdup(s);
+	if (dup_s == NULL)
+		return -FI_ENOMEM;
+
+	a = strtok_r(dup_s, ",", &saveptra);
+	while (a) {
+		first = last = -1;
+		stride = 1;
+		b = strtok_r(a, "-", &saveptrb);
+		assert(b);
+		first = atoi(b);
+		/* Check for range delimiter */
+		b = strtok_r(NULL, "-", &saveptrb);
+		if (b) {
+			c = strtok_r(b, ":", &saveptrc);
+			assert(c);
+			last = atoi(c);
+			/* Check for stride */
+			c = strtok_r(NULL, ":", &saveptrc);
+			if (c)
+				stride = atoi(c);
+		}
+
+		if (last == -1)
+			last = first;
+
+		for (j = first; j <= last; j += stride)
+			CPU_SET(j, &mycpuset);
+		a =  strtok_r(NULL, ",", &saveptra);
+	}
+
+	ret = pthread_setaffinity_np(mythread, sizeof(mycpuset), &mycpuset);
+	if (ret)
+		ret = -errno;
+
+	free(dup_s);
+	return ret;
+}
```

Comment:
So, I don't really like duplicating this function in two files given its complexity.  Maybe we need to move it to unix/osd.c instead and just use an ifdef to filter out osx.  What do you think?

Reply:
Yes, I agree. But OSX and Linux and FreeBSD use `unix/osd.c` and `unix/osd.h`
I think it would be more appropriate to move this code to the `unix/osd.c` file and move some places of OS-dep code of this function under `ifdef`

Reply:
Please, find the code updated as you proposed